src/livekit/openIDSFU.ts

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2023 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { MatrixClient } from "matrix-js-sdk";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import { Config } from "../config/Config";
+
+export interface SFUConfig {
+  url: string;
+  jwt: string;
+}
+
+export async function getSFUConfigWithOpenID(
+  client: MatrixClient,
+  roomName: string
+): Promise<SFUConfig> {
+  const openIdToken = await client.getOpenIdToken();
+  logger.debug("Got openID token", openIdToken);
+
+  const livekitCfg = Config.get().livekit;
+
+  if (!livekitCfg?.livekit_service_url) {
+    throw new Error("No livekit service URL defined");
+  }
+
+  const res = await fetch(livekitCfg.livekit_service_url + "/sfu/get", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      room: roomName,
+      openid_token: openIdToken,
+      remove_me_user_id: client.getUserId(), // the service will get this from the openid request
+      device_id: client.getDeviceId(),
+    }),
+  });
+  if (res.status / 100 !== 2) {
+    throw new Error("SFO Config fetch failed with status code " + res.status);
+  }
```

Comment:
Perhaps we could describe it a bit more explicitly, i.e. check for the "good" status and return the config to the user and fail otherwise, `status / 100 != 2` looks like too many magic numbers 🙂 (it's understandable that we do it based on the numeric numbers for different statuses, but I think explicitly written expectations are better and easier to refactor in the future should we use something different than HTTP).

What do you think?

Reply:
This is fairly standard in an HTTP client as HTTP defines anything in the 200-300 range as successful - it doesn't matter what the actual code is if you only care whether it worked or not. Checking for just one status would mean the client would break if the server changed to return a different 2xx code, which is perfectly valid. Would status >= 200 && status < 300 be any clearer? Or just a comment?

Reply:
Right, I understand that [a range from 200-299 is a successful response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#successful_responses), however, what I meant is that it's getting uncommon nowadays to rely on numeric client codes *in a high-level application code* (for the sake of the readability, being explicit / not needing to memorize the codes, examples in libraries: [one](https://www.npmjs.com/package/http-status-codes), [two](https://docs.rs/http/latest/http/status/struct.StatusCode.html#method.is_success)).

In this particular case, I think it would be more reliable (and arguably more readable) if we instead used the [`ok` property of a `Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response/status).

Also,

```typescript
if (res.status / 100 !== 2) {
    throw new Error("SFO Config fetch failed with status code " + res.status);
  }
```

will throw an exception in cases that we don't really expect since if the `status` is e.g. `202`, then `202 / 100` for a `number` will not be `2`, which will throw an exception.

Reply:
Oh, sure, we can use response.ok instead - done.